### PR TITLE
Added last seen to profile view

### LINKED ISSUE #2573: add Contact Profile View: last seen
sometime ago core added a last seen property to contacts, android displays that information in the profile screen of a contact.
how it looks on android:
image_2022-02-26_14-54-59

the android pr: https://github.com/deltachat/deltachat-android/pull/2144

### PR DESCRIPTION
**Pull request by canastillo**
For enhancement #2573 (closes #2573)

I'm having a problem with the timestamps. I'm getting dates from 1970. 

This is how it looks at the moment
image

Any suggestions?

### CODE REVIEW

**Comment by Simon-Laux** [comment]
> Any suggestions?

Screenshot shows 1970 date, better hide the text or say "long time ago" / "unknown" if timestamp is equal to `0`

Also we could discuss the location, like android does it in a section bellow the name, similar to the signature, but I don't have a strong opinion on that, maybe its fine as is (directly under the name)

**Comment by adbenitez** [comment]
> I'm having a problem with the timestamps. I'm getting dates from 1970.

hola Ana, gracias por contribuir,

dates from 1970 mean that you have never seen that contact in that case `Last seen: Unknown` must be shown (`last_seen_unknown` string)

**Comment by adbenitez** [comment]
oops! @Simon-Laux  already replied

**Comment by canastillo** [comment]
Thanks for your comments. I have made some changes for that case with the 70s. I wrote the condition as `<= 1970` because I remember seing even a date from 1969. 

I just have two questions: 
- Should I consider a later year or is 1970 okay?
- When does this "unknown" case happens? Because I chatted with the person of the screenshot like two days ago and I chatted today with another one and the same thing happens. Last seen has dates from 1970.

About the location, I'm open to suggestions. I thought that spot was the right place but I can change it. No problem.

**Comment by Jikstra** [comment]
> Thanks for your comments. I have made some changes for that case with the 70s. I wrote the condition as `<= 1970` because I remember seing even a date from 1969 (Dec 31st probably). I didn't write anything about `lastSeen == 0` 'cause I'm actually not getting 0s from the backend, but small numbers like 1653452517 or 1653437084.
> 
> I just have two questions:
> 
> * Should I consider a later year or is 1970 okay?
> * When does this "unknown" case happens? Because I chatted with the person of the screenshot like two days ago and I chatted today with another one and the same thing happens. Last seen has dates from 1970.
> 
> About the location, I'm open to suggestions. I thought that spot was the right place but I can change it. No problem.

I have the feeling it's something else, @adbenitez in theory 0 should be returned if there's no seen info right?

File: src/renderer/components/dialogs/ViewProfile.tsx [hunk 1]
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
...
+
   return (
     <div className='profile-info-name-container'>
       <div>
         <p className='group-name'>{name}</p>
       </div>
       <div className='address'>{address}</div>
+      <div>
+        <p>{`${lastSeenString}`}</p>
+      </div>
```
Comment (Jikstra) [inline src/renderer/components/dialogs/ViewProfile.tsx:61]:
```suggestion
      <div className='last-seen'>{lastSeenString}`}</div>
```

File: src/renderer/components/dialogs/ViewProfile.tsx [hunk 2]
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
...
+  const tx = useTranslationFunction()
+  const year = moment(lastSeen).format('YYYY')
+  let lastSeenString = ''
+
+  // Dates from 1970 mean that contact has never been seen
+  if (parseInt(year) <= 1970) lastSeenString = tx('last_seen_unknown')
+  else {
+    const extended = false
+    const date = formatRelativeTime(lastSeen, { extended })
+    lastSeenString = tx('last_seen_at', `${date}`)
```
Comment (Jikstra) [inline src/renderer/components/dialogs/ViewProfile.tsx:50]:
```suggestion
    lastSeenString = tx('last_seen_at', date)
```

File: src/renderer/components/dialogs/ViewProfile.tsx [hunk 3]
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
...
   address: string
+  lastSeen: number
 }) => {
+  const tx = useTranslationFunction()
+  const year = moment(lastSeen).format('YYYY')
+  let lastSeenString = ''
+
+  // Dates from 1970 mean that contact has never been seen
+  if (parseInt(year) <= 1970) lastSeenString = tx('last_seen_unknown')
+  else {
```
Comment (Jikstra) [inline src/renderer/components/dialogs/ViewProfile.tsx:47]:
```suggestion
  if (parseInt(year) <= 1970) {
    lastSeenString = tx('last_seen_unknown')
  } else {
```

File: src/renderer/components/dialogs/ViewProfile.tsx [hunk 4]
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
...
   name: string
   address: string
+  lastSeen: number
 }) => {
+  const tx = useTranslationFunction()
+  const year = moment(lastSeen).format('YYYY')
+  let lastSeenString = ''
+
+  // Dates from 1970 mean that contact has never been seen
+  if (parseInt(year) <= 1970) lastSeenString = tx('last_seen_unknown')
```
Comment (Jikstra) [inline src/renderer/components/dialogs/ViewProfile.tsx:46]:
```
  if (parseInt(year) <= 1970) {
    lastSeenString = tx('last_seen_unknown')
  } else {
```

**Review by Jikstra — COMMENTED** [review]
Added some small code changes :)

**Comment by Simon-Laux** [comment]
It turns out the timestamp from the core is in seconds, while the timestamp expected by momenjs is in milliseconds, so you need to do `lastSeen * 1000`:
const date = formatRelativeTime(lastSeen * 1000, { extended })

then you can also just compare to `0` instead of parsing the year. we also compare timestamps to `0` in other places, recently I added it for the chatlistitems of chats that have no messages yet.

After fixing it locally I noticed this:
image

which makes not much sense to me, so I think we should use another relative time format, so that it says "Monday" instead of just "Mon". Or we opt for just the absolute date/time? (has is it done on android @adbenitez ?)

File: src/renderer/components/dialogs/ViewProfile.tsx (hunk 4 above)
Comment (canastillo) [reply src/renderer/components/dialogs/ViewProfile.tsx:46]: Can you please write this as a suggestion so I can add it to the batch? The button doesn't appear in this one

**Comment by adbenitez** [comment]
> I have the feeling it's something else, @adbenitez in theory 0 should be returned if there's no seen info right?

yes, and that is what android checks for: 
https://github.com/deltachat/deltachat-android/blob/master/src/org/thoughtcrime/securesms/ProfileSettingsAdapter.java#L240

**Comment by adbenitez** [comment]
> which makes not much sense to me, so I think we should use another relative time format, so that it says "Monday" instead of just "Mon". Or we opt for just the absolute date/time? (has is it done on android @adbenitez ?)

for how it is done in android check:
https://github.com/deltachat/deltachat-android/blob/b488103c43fdd5c7dfa059cc1c6e33c1f0b813dd/src/org/thoughtcrime/securesms/util/DateUtils.java#L80

```
  public static String getExtendedTimeSpanString(final Context c, final Locale locale, final long timestamp) {
    StringBuilder format = new StringBuilder();
    if      (DateUtils.isToday(timestamp))                 {}
    else if (isWithin(timestamp,   6, TimeUnit.DAYS)) format.append("EEE ");
    else if (isWithin(timestamp, 365, TimeUnit.DAYS)) format.append("MMM d, ");
    else                                              format.append("MMM d, yyyy, ");

    if (DateFormat.is24HourFormat(c)) format.append("HH:mm");
    else                              format.append("hh:mm a");

    return getFormattedDateTime(timestamp, format.toString(), locale);
  }
```

File: src/renderer/components/dialogs/ViewProfile.tsx [hunk 5]
```diff
@@ -30,16 +32,31 @@ import { selectChat } from '../helpers/ChatMethods'
...
 }) => {
+  const tx = useTranslationFunction()
+  let lastSeenString = ''
+
+  // Dates from 1970 mean that contact has never been seen
+  if (lastSeen == 0) {
+    lastSeenString = tx('last_seen_unknown')
+  } else {
+    const extended = false
+    const date = formatRelativeTime(lastSeen * 1000, { extended })
```
Comment (Jikstra) [inline src/renderer/components/dialogs/ViewProfile.tsx:49]:
```suggestion
    const date = formatRelativeTime(lastSeen * 1000, { extended: false })
```

**Comment by canastillo** [comment]
Thanks @Simon-Laux for the explanation. I now understand what is going on. I changed that conditional and added the changes that Jikstra made. 

So, do you want me to add another relative format? I think on android they use the format that is used already in here.

File: src/renderer/components/dialogs/ViewProfile.tsx (hunk 5 above)
Comment (Jikstra) [reply src/renderer/components/dialogs/ViewProfile.tsx:49]: One last small code nitpick and then I'm happy to merge :P

**Comment by Jikstra** [comment]
> So, do you want me to add another relative format?

I wouldn't do it in this pr, but do we need another relative format? What was the discussion on this? Probably I missed it :P

File: src/renderer/components/dialogs/ViewProfile.tsx (hunk 5 above)
Comment (canastillo) [reply src/renderer/components/dialogs/ViewProfile.tsx]: Done! :D

**Pull request merged**